src/WebJobs.Script/ScriptHostBuilderExtensions.cs

```diff
@@ -476,5 +498,33 @@ private static void AddProcessRegistry(IServiceCollection services)
                 throw new InvalidOperationException($"The key '{key}' does not exist in the dictionary.");
             }
         }
+
+        private static void RecordAndThrowExternalStartupException(string message, Exception ex, ILoggerFactory loggerFactory)
+        {
+            var startupEx = new ExternalStartupException(message, ex);
+
+            var logger = loggerFactory.CreateLogger(LogCategories.Startup);
+            logger.LogDiagnosticEventError(DiagnosticEventConstants.ExternalStartupErrorCode, message, DiagnosticEventConstants.ExternalStartupErrorHelpLink, startupEx);
+
+            // Send the error to App Insights if possible. This is happening during ScriptHost construction so we
+            // have no existing TelemetryClient to use. Create a one-off client and flush it ASAP.
+            string appInsightsConnStr = Environment.GetEnvironmentVariable(EnvironmentSettingNames.AppInsightsConnectionString);
```

Comment:
I remember Jacob asking similar question before. Don't remember the answer. But should we use `IConfiguration` abstraction to read the env variable/app setting value?  
I recently chose to do that for a change I am working on instead of reading from Environment directly. Wondering I should change back to this approach.

Reply:
You almost always want to use `IConfiguration`. In some places we use the `IEnvironment` abstraction b/c we don't want to allow you to override some system env vars -- we want to go "straight to the source". But this case is an odd one b/c it's happening on startup so we don't have any DI container, and we may not have an `IConfiguration`.

There's 2 scenarios where this can get called b/c of an exception:
1. During `ConfigureAppConfiguration` -- at this point, I don't think we can trust `IConfiguration` as it hasn't been fully built. That's originally why I went straight to `Environment`.
2. During `ConfigureServices` -- at this point, at least, `IConfiguration` has been built, so we can use it. 

I made a change to do this -- at least this way, if someone is populating that connection string from KeyVault or something, they'll resolve correctly in `ConfigureServices`.

This is mostly a "best effort" logging attempt -- something's gone horribly wrong at startup and we just want to send it wherever we can.